File: 2023/python/src/aoc/solvers/d01.py
# -------------------------------------------------------------------------
#   Solver
# -------------------------------------------------------------------------

import re


# -------------------------------------------------------------------------
#   Part 1
# -------------------------------------------------------------------------


def part_1(inputs: str) -> int:
    total = 0
    for line in inputs.rstrip("\n").split("\n"):
        digits = [ch for ch in line if ch.isdigit()]
        total += int(digits[0] + digits[-1])

    return total


# -------------------------------------------------------------------------
#   Part 2
# -------------------------------------------------------------------------


def part_2(inputs: str) -> int:
    mapping = {
        "one": "1",
        "two": "2",
        "three": "3",
        "four": "4",
        "five": "5",
        "six": "6",
        "seven": "7",
        "eight": "8",
        "nine": "9",
    }
    pat = re.compile(f"(?=(\d|{'|'.join(mapping.keys())}))")

    total = 0
    for line in inputs.rstrip("\n").split("\n"):
        digits = [
            m if m.isnumeric() else mapping[m]
            for m in re.findall(pat, line)
        ]
        total += int(digits[0] + digits[-1])

    return total

File: 2023/cpp/src/solvers/d01.cpp
#include <map>
#include <regex>
#include <sstream>
#include <string>
#include <vector>
#include "d01.h"
#include "solution.h"

Solver01::Solver01() : Solution(1) {}

Solver01::~Solver01() {}

std::string Solver01::part_1(std::string inputs) {
  std::stringstream ss(inputs);
  std::string buffer;

  uint64_t total = 0;
  while (std::getline(ss, buffer, '\n')) {
    std::vector<std::string> found;
    for (auto ch : buffer) {
      int shift = ch - '0';
      if (shift >= 0 && shift < 10) {
        found.push_back(std::string(1, ch));
      }
    }
    std::string number = found.front() + found.back();
    total += std::stoi(number, NULL, 10);
  }

  return std::to_string(total);
}

std::string Solver01::part_2(std::string inputs) {
  std::stringstream ss(inputs);
  std::string buffer;

  std::regex redigits(
      "(?=(one|two|three|four|five|six|seven|eight|nine|zero|\\d))");
  std::map<std::string, int> mapping;
  mapping["one"] = 1;
  mapping["two"] = 2;
  mapping["three"] = 3;
  mapping["four"] = 4;
  mapping["five"] = 5;
  mapping["six"] = 6;
  mapping["seven"] = 7;
  mapping["eight"] = 8;
  mapping["nine"] = 9;
  mapping["zero"] = 0;
  mapping["1"] = 1;
  mapping["2"] = 2;
  mapping["3"] = 3;
  mapping["4"] = 4;
  mapping["5"] = 5;
  mapping["6"] = 6;
  mapping["7"] = 7;
  mapping["8"] = 8;
  mapping["9"] = 9;
  mapping["0"] = 0;

  uint64_t total = 0;
  while (std::getline(ss, buffer, '\n')) {
    std::vector<std::string> found;
    auto search_begin =
        std::sregex_iterator(buffer.begin(), buffer.end(), redigits);
    auto search_end = std::sregex_iterator();

    for (std::sregex_iterator i = search_begin; i != search_end; i++) {
      std::smatch match = *i;
      found.push_back(match.str(1));
    }

    total += 10 * mapping[found.front()] + mapping[found.back()];
  }

  return std::to_string(total);
}
